src/librustc_mir/interpret/machine.rs

```diff
@@ -221,17 +231,43 @@ pub trait Machine<'mir, 'tcx>: Sized {
         Ok(())
     }
 
-    /// Called for *every* memory access to determine the real ID of the given allocation.
-    /// This provides a way for the machine to "redirect" certain allocations as it sees fit.
+    /// Called for *every* memory access to determine the real ID of the given
+    /// allocation. This provides a way for the machine to "redirect" certain
+    /// allocations as it sees fit.
     ///
-    /// This is used by Miri to redirect extern statics to real allocations.
+    /// This is used by Miri for two purposes:
+    /// 1.  Redirecting extern statics to real allocations.
+    /// 2.  Creating unique allocation ids for thread locals.
```

Comment:
I worry that with the scheme you have set up currently, if you take a raw pointer to your thread local in one thread and send that pointer to another thread without modification, the other thread will access its own thread local instead of yours.

Reply:
This reminds me we still need to make borrowing a thread-local static an `Rvalue` of its own.

It's *fundamentally* a runtime operation, even if LLVM's representation of it is misleading.

Reply:
There's now #70685